### ISSUE TITLE
Ignore "unsafe method/property access" errors inside a corresponding `is` condition block

### ISSUE THREAD

**Issue by mphe** [godotengine/godot-proposals#8530]
### Describe the project you are working on

I enabled "Unsafe property/method access" warnings and set them to produce errors.
This helps immensely while porting Godot 3.x code to 4.x and it also prevents annoying errors during runtime that could have been easily catched in editor. 

### Describe the problem or limitation you are having in your project

When dealing with an object of a base type, e.g. the `InputEvent` variable received in `_input`, you might have to type check using `is` in order to act accordingly.
However, even if the condition matches, there are still errors about unsafe method/property access in the corresponding if block.

Example:

In the following, each `print()` line produces the error *"The property "axis" is not present on the inferred type "InputEvent" (but may be present on a subtype)."*, because the `event` variable does not in fact contain these properties.
Nevertheless, in this case it actually is a safe access.

```gdscript
func _input(event: InputEvent) -> void:
    if event is InputEventJoypadMotion:
        print(event.axis)  # Error
    elif event is InputEventJoypadButton:
        print(event.button_index)  # Error
    elif event is InputEventMouseMotion:
        print(event.relative)  # Error
    elif event is InputEventMouseButton or event is InputEventKey:
        print(event.button_index)  # Error

```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Unsafe property/method access should be considered safe when preceded by a corresponding `is` or `as` check.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The GDScript parser should figure out whether a variable has been explicitly type checked using `is` or `as` in the current scope and not throw errors when the correct type is ensured.
Code completion already works as expected and shows completions of the asserted type.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A workaround would be to use `as`.
The following shows an example with multiple ways to workaround it, however, all these workarounds are rather cumbersome and should be unnecessary.

```gdscript
func _input(event: InputEvent) -> void:
    var joypad_motion := event as InputEventMouseMotion

    if event is InputEventJoypadMotion:
        print((event as InputEventJoypadMotion).axis)
    elif event is InputEventJoypadButton:
        var ev := event as InputEventJoypadButton
        print(ev.button_index)
    elif joypad_motion:
        print(joypad_motion.relative)
    # ...
```

### Is there a reason why this should be core and not an add-on in the asset library?

It's GDScript parser/compiler related.

**Comment by dalexeev** [comment]
What you are asking for involves more advanced static analysis (implicit type change, accounting for assignments and other operations that can change the value/type at runtime). An alternative is a syntax for explicit type narrowing.

* See also #6941.

**Comment by kraftwer1** [comment]
What I don't understand is that everything is a "Node", even though Godot could know better. For example:

    $Enemy.position

…throws…

    gdscript: (UNSAFE_PROPERTY_ACCESS): The property "position" is not present on the inferred type "Node" (but may be present on a subtype).

However, `$Enemy` is of type `MeshInstance3D` – it's defined like that in the editor.

This makes the whole `UNSAFE_PROPERTY_ACCESS` pretty useless for me, although it could be such a helpful feature.

**Comment by dalexeev** [comment]
> However, `$Enemy` is of type `MeshInstance3D` – it's defined like that in the editor.

`$Enemy` is of type `Node`. `$NodePath` is a shorthand for `get_node(^"NodePath")`. `get_node()` has static return type `Node`. The fact that the editor is able to determine the node type for the currently open scene and provide more appropriate autocompletion suggestions does not affect the static analyzer. At runtime, a node can be removed and replaced with a node of a different type.

**Comment by kraftwer1** [comment]
Thanks for the clarification. Unfortunately, having hundreds of „as“ throughout the code makes it not only noisy, but also hard to maintain (imagine changing the type of a node). One of the main selling points of GDScript is its tight integrarion into the engine, and this could exactly be an opportunity: $Enemy being more than just a shorthand. It could be something that its known by the static analyzer and the static analyzer could be less generic than let‘s say C# or any other truly foreign language.

**Comment by geekley** [comment]
IMO the best option for explicit type narrowing (aside from `match`) would be to go the C# route and allow `is` to specify the narrowing variable directly. So the `is` expression returns a bool as normal, but also declares a variable that is in scope in the respective branch (e.g. inside the if when checking a positive match, or in the else branch when checking "is not").

I suggest the same familiar syntax as C#: `x is T t`. The minor syntax difference that the variable is after the type (which is a bit different from GDScript order `t: T`) shouldn't be an issue in this special case -- it works, it's succinct and we can't use an extra colon anyway as it would clash with the "if" colon.

```gdscript
func _input(event: InputEvent) -> void:
    if event is InputEventJoypadMotion e:
        print(e.axis) # variable `e` is in scope here
    elif event is InputEventJoypadButton e: # previous `e` was not in scope here, so we can redeclare it
        print(e.button_index)
    elif event is InputEventMouseMotion e:
        print(e.relative)
    # this one is only if union types is ever implemented:
    elif event is (InputEventMouseButton | InputEventKey) e:
        print(e.button_index)

    # short-circuit and/or should also count as logical branches for scope analysis:
    if $Node is CollisionShape2D shape and shape.disabled:
        print(shape.disabled)
```

**Comment by Zshandi** [comment]
Is there an issue yet specifically for adding the implicit static analyzer type narrowing in general (as in for if and other conditionals such as match, as well as including autocomplete and not just removing the warning)? And if not, should we create a new one or just update this one to encompass that?

**Comment by dalexeev** [comment]
> Is there an issue yet specifically for adding the implicit static analyzer type narrowing in general

See https://github.com/godotengine/godot-proposals/issues/10749#issuecomment-3196916086.

The problem is that it is not that simple. It requires more advanced static analysis, in particular control flow analysis. Probably, it requires the introduction of a new data structure: the control flow graph (CFG), since now the compiler directly transforms the abstract syntax tree (AST) into bytecode.

**Comment by PureAsbestos** [comment]
This presents itself like a bug, since in the simple case of
```gdscript
func _input(event: InputEvent) -> void:
    if event is InputEventJoypadMotion:
        print(event.axis)
```
you get perfectly good behavior and autocomplete with the warning set to Ignore.

However, with only slightly more complexity, autocomplete will fail as well:
```gdscript
func _input(event: InputEvent) -> void:
    if event is InputEventJoypadMotion and true:
        print(event.axis)
```

This demonstrates why it is not easy to fix, the autocomplete is already just getting around this with a hack. Ideally both autocomplete and type checking will do a proper analysis in the future.

Maybe porting the hack to the type checker as a temporary fix isn't such a bad idea though? Not sure.

**Comment by Copperplate** [comment]
> The GDScript parser should figure out whether a variable has been explicitly type checked using is or as in the current scope and not throw errors when the correct type is ensured.

Yes, this would be a super-valuable feature to have, and make this far more intuitive!
